### ISSUE TITLE
dri3 and glamor

### ISSUE THREAD

**Issue by jsorg71** [neutrinolabs/xrdp#1029]
With DRI3, we can enable hardware OpenGL in xrdp sessions.  DRI3 allows non root access to video hardware through render nodes(/dev/dri/renderD128).
With Glamor, we can use that hardware OpenGL to accelerate the 2D calls with pixels shaders.

**Comment by jsorg71** [comment]
Here is a working prototype.  This is Windows 10 mstsc connected to an xrdp ATI GPU server.
Showing the KDE desktop cube affect.
![xrdp-glamor](https://user-images.githubusercontent.com/666160/35781844-2eb6974e-09e7-11e8-91b2-01b22892ecb8.png)

**Comment by paravz** [comment]
do you need help, maybe in testing this ?

**Comment by jsorg71** [comment]
Sure, you can use glamor branch
https://github.com/jsorg71/xorgxrdp/tree/glamor
Make sure your Xorg is 1.19+ (Debian 9 or Ubuntu 18.04)
Should work with NVidia, ATI and Intel graphics as long as the open source drivers are used.
No change to xrdp yet, this is just changes to the Xorg driver.

**Comment by paravz** [comment]
Finally got around to test on fedora 28 with cinnamon. Basically works great on the first glance. As expected, significant reduction in CPU usage across the board - Thanks for this change!

The only issue connecting from win10 mstsc was blank screen depending on client's "connection type":
* my client for xorgxrd-standard+Cinnamon with software rendering worked with: "connection type:i:7" (auto detect network settings). 
* Connecting to xorgxrdp-glamor with "connection type:i:7" results in black/blank screen in mstsc.
* xorgxrdp-glamor+Cinnamon DRI3 requires "connection type:i:6" (LAN 10 Mbps). 
* xorgxrdp+Cinnamon software rendering works with "connection type:i:6" too. Basically i switched to "connection type:i:6" for xrdp systems

**Comment by Nexarian** [comment]
Looks like the glamor branch has been merged into xorgxrdp! That was a lot of work!

Cam this work with Nvidia? I see provisions in the code for EGL streams (Nvidia's tech), but when I tried it, it's still using llvmpipe (software rendering) on Nvidia's binary drivers (Ubuntu 18.04)

**Comment by jsorg71** [comment]
Yes, it's merge in devel branch but off by default.  --enable-glamor to turn it on.
The glamor module in Xorg only uses mesa currently.  So it works with the nouveau driver on NVidia hardware.  I know that's not optimal but we'll work on a solution.  Currently have to recommend Intel or AMD hardware.

**Comment by Nexarian** [comment]
@jsorg71 That's great that you're working on a solution! Having XRDP work with the Nvidia accelerated ecosystem would unblock remoting scenarios in the Machine Learning Linux world (which is huge).

I am happy to test prototypes when the come out, and I can verify that it is perfectly stable with glamor enabled with Nvidia drivers, it just uses software rendering (which is important, too).

What does libepoxy and the EGL stream portion do?

**Comment by jsorg71** [comment]
I see this link about 'PRIME render offloading' that looks interesting.  This is all really new and relies on xserver bits not yet released.

libepoxy is just a helper library for OpenGL.
https://github.com/anholt/libepoxy
rdpEGL.c is the custom pixel shades to do the RGB to YUV, CRC32, and other operations on the GPU.

**Comment by bernstein82** [comment]
--enable-glamor doesn't seem to be present in ubuntu 20.04 xrdp (release 0.9.12).

compiled from source... but there's still no --enable-glamor option... have things changed?

also: does it work with xwayland or do i need x11?

**Comment by rkkoszewski** [comment]
Where do you put the **--enable-glamor** flag?

**Comment by paravz** [comment]
to configure script of xorgxrdp, when building from source

**Comment by steverweber** [comment]
Sorry if not helpful.. A week or so ago I went to rebuild parts of the xrdp stack to test glamor on ubuntu20.04(pre release).. I then stopped after seeing the xorgxrdp package seemed to include glamoregl in: `/etc/X11/xrdp/xorg.conf`. After making an xrdp connection and running glxgears (seemed ok).. but that might just be software rendering seeing llvm `glxinfo grep | llvmpipe`.

Shrug some directions on how to get testing this would be helpful. I help manage some HPCs ~50 nvidia gpus. Would love to give the users hw accelerated glxgears over rdp.

**Comment by rkkoszewski** [comment]
Nice! It works! I can see the GPU usage in intel_gpu_top, and it seems to leverage the GPU really nicely (You can even watch videos and run graphical tasks), while the CPU is completely relaxed (It barely uses 3-5% of CPU on average). 

But the frame-rate seems extremely low in XRDP. It looks like it's running at 5 to 10 FPS or so, no matter if moving windows around or watching a video. I had the same issues in software rendering mode previously. Is that normal? (I'm in LAN, tested with Experience "LAN 10Mbps" and all lower options, have set the minimum security to None in XRDP, and also tried with 16bit colors, all with the same result).

**Comment by paravz** [comment]
@rkkoszewski try changing xrdp.ini settings, specifically bitmap_compression bulk_compression and independently try "Detect automatically" instead of "LAN 10Mbs" in client settings

Disabling all compression and using "Detect automatically" should be best quality, with enormous bandwidth requirements - but i'm curious what will be your experience

**Comment by rkkoszewski** [comment]
@paravz Thank you for the suggestions. I tested every possible combination of bitmap_compression, bulk_compression  and the "Detect automatically"/"LAN 10Mbs", but the results are nearly always the same. It does look that enabling bulk_compression makes it sligthly slower, but it might be placebo. The only difference was in bandwith usage. It still looks like it's running at a static 5 to 10 FPS.

One thing to note is that I'm running the desktop in an LXC container (Glamor is 100% working, as I can see the GPU load change when dragging windows). Tested also with Windows Native RDP client, as also with the open source aRDP client, all with same result. Setting the resolution to 800x600 seems to improve things (Both compressed and uncompressed), but again might be placebo, as the difference is minimal. 

Any more ideas? How is it working for you? Is it possible to watch smoothly a video (even if it's just in a small window) trough xrdp?

**Comment by paravz** [comment]
1080p youtube video in firefox works smoothly for me in cinnamon, it might your hardware, kernel or something else

**Comment by seamus-x** [comment]
Seems to have broken xrdp after I rebuilt xorgxrdp with --enable-glamorx

[20200425-14:25:38] [INFO ] Socket 12: AF_INET connection received from 10.0.0.123 port 46380
[20200425-14:25:38] [DEBUG] Closed socket 12 (AF_INET 10.0.0.100:3389)
[20200425-14:25:38] [DEBUG] Closed socket 11 (AF_INET 0.0.0.0:3389)
[20200425-14:25:38] [INFO ] Using default X.509 certificate: /etc/xrdp/cert.pem
[20200425-14:25:38] [INFO ] Using default X.509 key file: /etc/xrdp/key.pem
[20200425-14:25:38] [DEBUG] TLSv1.3 enabled
[20200425-14:25:38] [DEBUG] TLSv1.2 enabled
[20200425-14:25:38] [DEBUG] Security layer: requested 3, selected 1
[20200425-14:25:38] [INFO ] connected client computer name: Chrome
[20200425-14:25:38] [INFO ] adding channel item name rdpdr chan_id 1004 flags 0xc0800000
[20200425-14:25:38] [INFO ] adding channel item name rdpsnd chan_id 1005 flags 0xc0000000
[20200425-14:25:38] [INFO ] adding channel item name cliprdr chan_id 1006 flags 0xc0a00000
[20200425-14:25:38] [INFO ] adding channel item name drdynvc chan_id 1007 flags 0xc0800000
[20200425-14:25:38] [INFO ] TLS connection established from 10.0.0.123 port 46380: TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384
[20200425-14:25:38] [DEBUG] xrdp_00002639_wm_login_mode_event_00000001
[20200425-14:25:38] [INFO ] Loading keymap file /etc/xrdp/km-00000409.ini
[20200425-14:25:38] [WARN ] local keymap file for 0x00000409 found and doesn't match built in keymap, using local keymap file
[20200425-14:25:38] [DEBUG] xrdp_wm_log_msg: connecting to sesman ip 127.0.0.1 port 3350
[20200425-14:25:38] [INFO ] xrdp_wm_log_msg: sesman connect ok
[20200425-14:25:38] [DEBUG] xrdp_wm_log_msg: sending login info to session manager, please wait...
[20200425-14:25:38] [DEBUG] return value from xrdp_mm_connect 0
[20200425-14:25:38] [INFO ] xrdp_wm_log_msg: login successful for display 10
[20200425-14:25:38] [DEBUG] xrdp_wm_log_msg: started connecting
[20200425-14:25:41] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:25:45] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:25:48] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:25:52] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:25:55] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:25:59] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:26:02] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:26:05] [DEBUG] Closed socket 11 (AF_INET 0.0.0.0:3389)
[20200425-14:26:06] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:26:06] [DEBUG] xrdp_wm_log_msg: some problem
[20200425-14:26:06] [DEBUG] xrdp_mm_module_cleanup
[20200425-14:26:06] [DEBUG] Closed socket 18 (AF_INET 127.0.0.1:42254)
[20200425-14:26:06] [DEBUG] Closed socket 12 (AF_INET 10.0.0.100:3389)
[20200425-14:26:06] [DEBUG] xrdp_mm_module_cleanup
[20200425-14:26:06] [INFO ] address [0.0.0.0] port [3389] mode 1
[20200425-14:26:06] [INFO ] listening to port 3389 on 0.0.0.0
[20200425-14:26:06] [INFO ] xrdp_listen_pp done
[20200425-14:26:06] [DEBUG] Closed socket 7 (AF_INET 0.0.0.0:3389)
[20200425-14:26:08] [INFO ] starting xrdp with pid 9808
[20200425-14:26:08] [INFO ] address [0.0.0.0] port [3389] mode 1
[20200425-14:26:08] [INFO ] listening to port 3389 on 0.0.0.0
[20200425-14:26:08] [INFO ] xrdp_listen_pp done
[20200425-14:26:26] [DEBUG] Closed socket 11 (AF_INET 0.0.0.0:3389)
[20200425-14:26:26] [INFO ] address [0.0.0.0] port [3389] mode 1
[20200425-14:26:26] [INFO ] listening to port 3389 on 0.0.0.0
[20200425-14:26:26] [INFO ] xrdp_listen_pp done
[20200425-14:26:26] [DEBUG] Closed socket 7 (AF_INET 0.0.0.0:3389)
[20200425-14:26:28] [INFO ] starting xrdp with pid 9843
[20200425-14:26:28] [INFO ] address [0.0.0.0] port [3389] mode 1
[20200425-14:26:28] [INFO ] listening to port 3389 on 0.0.0.0
[20200425-14:26:28] [INFO ] xrdp_listen_pp done
[20200425-14:26:42] [INFO ] Socket 12: AF_INET connection received from 10.0.0.123 port 46390
[20200425-14:26:42] [DEBUG] Closed socket 12 (AF_INET 10.0.0.100:3389)
[20200425-14:26:42] [DEBUG] Closed socket 11 (AF_INET 0.0.0.0:3389)
[20200425-14:26:42] [INFO ] Using default X.509 certificate: /etc/xrdp/cert.pem
[20200425-14:26:42] [INFO ] Using default X.509 key file: /etc/xrdp/key.pem
[20200425-14:26:42] [DEBUG] TLSv1.3 enabled
[20200425-14:26:42] [DEBUG] TLSv1.2 enabled
[20200425-14:26:42] [DEBUG] Security layer: requested 3, selected 1
[20200425-14:26:42] [INFO ] connected client computer name: Chrome
[20200425-14:26:42] [INFO ] adding channel item name rdpdr chan_id 1004 flags 0xc0800000
[20200425-14:26:42] [INFO ] adding channel item name rdpsnd chan_id 1005 flags 0xc0000000
[20200425-14:26:42] [INFO ] adding channel item name cliprdr chan_id 1006 flags 0xc0a00000
[20200425-14:26:42] [INFO ] adding channel item name drdynvc chan_id 1007 flags 0xc0800000
[20200425-14:26:42] [INFO ] TLS connection established from 10.0.0.123 port 46390: TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384
[20200425-14:26:42] [DEBUG] xrdp_0000267a_wm_login_mode_event_00000001
[20200425-14:26:42] [INFO ] Loading keymap file /etc/xrdp/km-00000409.ini
[20200425-14:26:42] [WARN ] local keymap file for 0x00000409 found and doesn't match built in keymap, using local keymap file
[20200425-14:26:42] [DEBUG] xrdp_wm_log_msg: connecting to sesman ip 127.0.0.1 port 3350
[20200425-14:26:42] [INFO ] xrdp_wm_log_msg: sesman connect ok
[20200425-14:26:42] [DEBUG] xrdp_wm_log_msg: sending login info to session manager, please wait...
[20200425-14:26:42] [DEBUG] return value from xrdp_mm_connect 0
[20200425-14:26:42] [INFO ] xrdp_wm_log_msg: login successful for display 10
[20200425-14:26:42] [DEBUG] xrdp_wm_log_msg: started connecting
[20200425-14:26:45] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:26:49] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:26:52] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:26:56] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:26:59] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:27:03] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:27:06] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:27:10] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:27:13] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:27:17] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:27:20] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:27:24] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:27:27] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:27:31] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:27:34] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:27:38] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:27:41] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:27:45] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:27:48] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:27:52] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:27:55] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:27:59] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:28:02] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:28:06] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:28:09] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:28:13] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:28:16] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:28:20] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:28:23] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:28:27] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:28:31] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:28:34] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:28:38] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:28:41] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:28:45] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:28:48] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:28:52] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:28:55] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:28:59] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:29:02] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:29:06] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:29:09] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:29:13] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:29:16] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:29:20] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:29:23] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:29:27] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:29:30] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:29:34] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:29:37] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:29:41] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:29:44] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:29:48] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:29:51] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:29:55] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:29:58] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:30:02] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:30:05] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:30:09] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:30:12] [DEBUG] xrdp_wm_log_msg: connection problem, giving up
[20200425-14:30:12] [DEBUG] Closed socket 19 (AF_UNIX)
[20200425-14:30:12] [DEBUG] xrdp_wm_log_msg: some problem
[20200425-14:30:12] [DEBUG] xrdp_mm_module_cleanup
[20200425-14:30:12] [DEBUG] Closed socket 18 (AF_INET 127.0.0.1:42256)
[20200425-16:11:34] [DEBUG] Closed socket 12 (AF_INET 10.0.0.100:3389)
[20200425-16:11:34] [DEBUG] xrdp_mm_module_cleanup

**Comment by paravz** [comment]
Interestingly my system uses different "X" drivers with i915 hardware, especially VDPAU for hardware video acceleration. I have not specifically tuned any of it, comes with 5.4.19 kernel
```
[925218.683] rdpDri2Init:
[925218.683] (II) XRDPDEV(0): [DRI2] Setup complete
[925218.683] (II) XRDPDEV(0): [DRI2]   DRI driver: i965
[925218.683] (II) XRDPDEV(0): [DRI2]   VDPAU driver: va_gl

```

**Comment by seamus-x** [comment]
@rkkoszewski Have you tried recompiling xrdp with --enable-rfxcodec and enable RemoteFX on the client?

**Comment by rkkoszewski** [comment]
@seamus-x I thought that option was enabled by default. Is there any way to check that it is indeed missing in my setup? I can't see anything rfx related in the logs I posted. Should it be there? Or somewhere else?

**Comment by 2X3** [comment]
Unfortunately this does not work for me on Manjaro. Is it possible that with `--enable-glamor` option different version of `xorg-server` is expected?

### Errors in Xorg log
```
[1030848.611] (II) LoadModule: "xrdpdev"
[1030848.611] (II) Loading /usr/lib/xorg/modules/drivers/xrdpdev_drv.so
[1030848.611] (EE) Failed to load /usr/lib/xorg/modules/drivers/xrdpdev_drv.so: /usr/lib/xorg/modules/drivers/xrdpdev_drv.so: undefined symbol: rdpRRModeDestroy
[1030848.611] (EE) Failed to load module "xrdpdev" (loader failed, 0)
[1030848.611] (II) LoadModule: "xrdpmouse"
[1030848.611] (II) Loading /usr/lib/xorg/modules/input/xrdpmouse_drv.so
[1030848.611] (EE) Failed to load /usr/lib/xorg/modules/input/xrdpmouse_drv.so: /usr/lib/xorg/modules/input/xrdpmouse_drv.so: undefined symbol: rdpUnregisterInputCallback
[1030848.611] (EE) Failed to load module "xrdpmouse" (loader failed, 0)
[1030848.611] (II) LoadModule: "xrdpkeyb"
[1030848.611] (II) Loading /usr/lib/xorg/modules/input/xrdpkeyb_drv.so
[1030848.611] (EE) Failed to load /usr/lib/xorg/modules/input/xrdpkeyb_drv.so: /usr/lib/xorg/modules/input/xrdpkeyb_drv.so: undefined symbol: rdpUnregisterInputCallback
[1030848.611] (EE) Failed to load module "xrdpkeyb" (loader failed, 0)
[1030848.611] (EE) No drivers available.
[1030848.611] (EE) 
Fatal server error:
[1030848.611] (EE) no screens found(EE) 
```

### Package versions
```
pacman -Q xrdp xorgxrdp-git xorg-server
xrdp 0.9.13-1
xorgxrdp-git 0.2.13-1
xorg-server 1.20.8-3
```

**Comment by lixaotec** [comment]
finally i managed to get it.. 
What is to be to to have NVidia Proprietary driver to works with it?

**Comment by Nexarian** [comment]
Nvidia is a bit tricky. @jsorg71 created a prototype branch that allows for direct hardware acceleration with Nvidia: https://github.com/jsorg71/xorgxrdp/tree/nvidia_hack

But it requires some setup:
1. I couldn't get this to work with Gnome. I recommend using mate.
2. You'll need to rebuild xorgxrdp and install it first.
3. Find what PCI address your Nvidia card is on:
`lspci | grep NVIDIA`
It’ll look something like this:
`65:00.0 VGA compatible controller: NVIDIA Corporation GP104GL [Quadro P5000] (rev a1)`
That 65 value is the PCI bus address of your card.
Convert that value to hexadecimal (easy to do with a Google search)
4. Now, edit the new configuration file for the Nvidia specific setup: sudo vim /etc/X11/xrdp/xorg_nvidia.conf
5. Change Line 38 to replace the ‘2’ with whatever hex value you obtained above
6. Now, open up /etc/xrdp/sesman.ini and change the line that has param=xrdp/xrdp.conf to param=xorg_nvidia.conf — On my system this line is 89 but YMMV
7. Restart xrdp: `sudo service xrdp restart` 

Confirm it works by running glxgears.

Also note that this doesn't use virtualgl, it's directly integrated with the drivers!

Right now this is just a prototype and obviously the awkward setup and other issues such as Gnome compatibility need to be addressed, but it's a massive step in the right direction!

**Comment by lixaotec** [comment]
Thanks @Nexarian ! Will soon try and let all know.

Also would like to contribute with coding, despite the fact i dont know nothing on this, if manteiners could provide and describs what else is to be done and where to start i could try.

I think this is the only open tech nowadays that enables serius graphics remote working, without goingo to m$ world or online services

**Comment by rkkoszewski** [comment]
If somebody is struggling to install XRDP with GLAMOR enabled here are my notes on how to install it on a Debian 9: https://gist.github.com/rkkoszewski/aa531cee7126edf329b76bdd0546f502

A question for @jsorg71 : Is there any way to make VAAPI with Hardware acceleration to work in a XRDP session with GLAMOR enabled? From my tests and by using "strace" I can see that VAAPI is working, it is able to open the GPU render node, but then it checks for some "MAGIC" number that is different from the one it is expecting, and it defaults to software encoding/decoding. (I'll upload the log once I find it again)

**Comment by lixaotec** [comment]
> If somebody is struggling to install XRDP with GLAMOR enabled here are my notes on how to install it on a Debian 9: https://gist.github.com/rkkoszewski/aa531cee7126edf329b76bdd0546f502
> 
> A question for @jsorg71 : Is there any way to make VAAPI with Hardware acceleration to work in a XRDP session with GLAMOR enabled? From my tests and by using "strace" I can see that VAAPI is working, it is able to open the GPU render node, but then it checks for some "MAGIC" number that is different from the one it is expecting, and it defaults to software encoding/decoding. (I'll upload the log once I find it again)

Is it thr same for Cinnamon (mint)

**Comment by sshaikh** [comment]
I'm having difficulty getting this running in Arch.

It compiles okay but then sesman throws an error when trying to create an x server:

`
[ERROR] X server for display 10 startup timeout
[CORE ] waiting for window manager (pid 804) to exit
[ERROR] X server for display 10 startup timeout
[ERROR] another Xserver might already be active on display 10 - see log
[CORE ] window manager (pid 804) did exit, cleaning up session
`
Any tips on how to debug this?

**Comment by sshaikh** [comment]
I see similar errors here:

https://github.com/neutrinolabs/xrdp/issues/1411
https://github.com/neutrinolabs/xrdp/issues/786

So it seems a bit of a hairy one and is now above my paygrade unfortunately.

**Comment by MakiseKurisu** [comment]
glamor seems great and I really want to see this feature is enabled by default, or configurable at run time, not build time.

**Comment by jat001** [comment]
> I see similar errors here:
> 
> #1411
> #786
> 
> So it seems a bit of a hairy one and is now above my paygrade unfortunately.
> 
> Edit: fixed here: https://bbs.archlinux.org/viewtopic.php?pid=1937867#p1937867 by using the following configure:
> 
> ```
> CFLAGS+=" $(pkgconf --cflags libdrm)" LDFLAGS+=" -L/usr/lib/xorg/modules -lglamoregl -Wl,-rpath,/usr/lib/xorg/modules" ./configure \
>     --prefix="/usr" -enable-glamor
> ```
> 
> Can this be used to fix something in the source?

After I made a package for [xorgxrdp with glamor](https://aur.archlinux.org/packages/xorgxrdp-glamor/), I realized it [does not work on nvidia gpu](https://github.com/neutrinolabs/xrdp/issues/1690#issuecomment-696892738). however, feel free to test it.

**Comment by jat001** [comment]
> > I see similar errors here:
> > #1411
> > #786
> > So it seems a bit of a hairy one and is now above my paygrade unfortunately.
> > Edit: fixed here: https://bbs.archlinux.org/viewtopic.php?pid=1937867#p1937867 by using the following configure:
> > ```
> > CFLAGS+=" $(pkgconf --cflags libdrm)" LDFLAGS+=" -L/usr/lib/xorg/modules -lglamoregl -Wl,-rpath,/usr/lib/xorg/modules" ./configure \
> >     --prefix="/usr" -enable-glamor
> > ```
> > 
> > 
> > Can this be used to fix something in the source?
> 
> After I made a package for [xorgxrdp with glamor](https://aur.archlinux.org/packages/xorgxrdp-glamor/), I realized it [does not work on nvidia gpu](https://github.com/neutrinolabs/xrdp/issues/1690#issuecomment-696892738). however, feel free to test it.

I also made a package for [xorgxrdp with nvidia](https://aur.archlinux.org/packages/xorgxrdp-nvidia-git/) and I can confirm that it works.

**Comment by sshaikh** [comment]
@jat001 Thanks for this. I've passed some feedback on the aur page for the non nvidia PKGBUILD

**Comment by rcarmo** [comment]
I've been trying to get this to work but once I've logged in, the screen remains with the `sesman` wallpaper and there is nothing in the logs - except a timeout, eventually. 

Although I don't think that's necessary, I checked that my user is part of the `video` group (so I can access `/dev/dri`).

Anything obvious to look for in the logs, or debugging methods for the `.so`? If I revert to my distro's `xorgxrdp` package everything works fine.

**Comment by rcarmo** [comment]
I've been trying to get this to work but once I've logged in, the screen remains with the `sesman` wallpaper and there is nothing in the logs - except a timeout, eventually:

```
[20221018-17:33:29] [INFO ] sesman connect ok
[20221018-17:33:29] [INFO ] sending login info to session manager. Please wait...
[20221018-17:33:29] [INFO ] xrdp_wm_log_msg: login successful for user me on display 10
[20221018-17:33:29] [INFO ] login successful for user me on display 10
[20221018-17:33:29] [INFO ] loaded module 'libxup.so' ok, interface size 10296, version 4
[20221018-17:33:29] [INFO ] started connecting
[20221018-17:33:29] [INFO ] lib_mod_connect: connecting via UNIX socket
```

Although I don't think that's necessary, I checked that my user is part of the `video` group (so I can access `/dev/dri`).

Anything obvious to look for in the logs, or debugging methods for the `.so`? If I revert to my distro's `xorgxrdp` package everything works fine.

**Comment by paravz** [comment]
Check instructions from https://github.com/neutrinolabs/xrdp/issues/1029#issuecomment-679958758, including X11 setup.

Check your glamor device name and if it's configured and working, ie:

```
$ lsgpu 
card0                    xxxx:xxxx                         drm:/dev/dri/card0
└─renderD128                                               drm:/dev/dri/renderD128

$ cat /etc/X11/xrdp/xorg.conf |grep render
    Option "DRMDevice" "/dev/dri/renderD128"

$ sudo journalctl -l --all -b |grep renderD128
Oct 18 22:45:42 dev xrdp-sesman[130071]: rdpProbe: found DRMDevice xorg.conf value [/dev/dri/renderD128]
Oct 18 22:45:42 dev xrdp-sesman[130071]: rdpPreInit: /dev/dri/renderD128 open ok, fd 12

```

**Comment by DocMAX** [comment]
Can this be run headless? I mean with something like the Xvnc backend?

**Comment by rcarmo** [comment]
It does run headless with the Xrdp backend. I have been using `xorgxrdp-glamor` (and equivalent packages shipping in Fedora, Ubuntu, etc.) for a long while now to get GPU-accelerated remote desktops. Sometimes it breaks (because of re-packaging issues, since not many people know about glamor, RFX, etc.), but when things are packaged "right", it is a very smooth experience. I only wish there was better debugging output (not just for `glamor` not working, but also including negotiated codecs).

**Comment by DocMAX** [comment]
Testing the on Arch with AUR package xorgxrdp-glamor now.

**Comment by rcarmo** [comment]
Make sure you edit `/etc/xrdp/xrdp.ini` to uncomment this section:

```ini
[Xorg]
name=Xorg
lib=libxup.so
username=ask
password=ask
ip=127.0.0.1
port=-1
code=20
```

(or equivalent) and make sure you select Xorg to login (or just disable VNC altogether). Also, client settings matter:

```bash
xfreerdp /gfx:rfx +gfx-progressive  /sound /dynamic-resolution /v:fedora.lan
```

**Comment by DocMAX** [comment]
Awesome!!! It's working.
I was working with Xvnc before. Now i have fully GPU accelerated RDP. If only I had come across it earlier.

**Comment by rcarmo** [comment]
I honestly don't know why `Xrdp` isn't the default, since it is faster than Xvnc even without `glamor`...

**Comment by DocMAX** [comment]
Actually i noticed Xrdp is slower than Xvnc to me (opening Firefox for example takes longer, no idea why). But i stay on Xorg to have 3D.

**Comment by matt335672** [comment]
I think this can now be closed as we've made it easier to use `--enable-glamor`.

See neutrinolabs/xorgxrdp#295, which is largely the work of @jat001.
